src/TreeSitter/Language.hs

```diff
@@ -1,54 +1,45 @@
+
+import           Data.Ix (Ix)
+import qualified Data.Set as Set
+import           Data.Traversable (for)
+import           Data.Word
+import           Foreign.C.String
+import           Foreign.Ptr
+import           Language.Haskell.TH
+import           Language.Haskell.TH.Syntax
+import           System.Directory
+import           System.FilePath.Posix
+import           TreeSitter.Symbol
 
 newtype Language = Language ()
   deriving (Show, Eq)
 
-type TSSymbol = Word16
-
-data SymbolType = Regular | Anonymous | Auxiliary
-  deriving (Enum, Eq, Ord, Show)
-
 foreign import ccall unsafe "ts_language_symbol_count" ts_language_symbol_count :: Ptr Language -> Word32
 foreign import ccall unsafe "ts_language_symbol_name" ts_language_symbol_name :: Ptr Language -> TSSymbol -> CString
 foreign import ccall unsafe "ts_language_symbol_type" ts_language_symbol_type :: Ptr Language -> TSSymbol -> Int
 
-
-class (Bounded s, Enum s, Ix s, Ord s, Show s) => Symbol s where
-  symbolType :: s -> SymbolType
-
-
 -- | TemplateHaskell construction of a datatype for the referenced Language.
 mkSymbolDatatype :: Name -> Ptr Language -> Q [Dec]
 mkSymbolDatatype name language = do
   symbols <- (++ [(Regular, "ParseError")]) <$> runIO (languageSymbols language)
-  let namedSymbols = renameDups [] $ uncurry symbolToName <$> symbols
+  let namedSymbols = renameDups $ ((,) . fst <*> uncurry symbolToName) <$> symbols
 
   Module _ modName <- thisModule
-  pure
-    [ DataD [] name [] Nothing (flip NormalC [] . mkName . snd <$> namedSymbols) [ DerivClause Nothing [ ConT ''Show, ConT ''Enum, ConT ''Eq, ConT ''Ord, ConT ''Bounded, ConT ''Ix ] ]
-    , InstanceD Nothing [] (AppT (ConT ''Symbol) (ConT name)) [ FunD 'symbolType (uncurry (clause modName) <$> namedSymbols) ] ]
-  where clause modName symbolType str = Clause [ ConP (Name (OccName str) (NameQ modName)) [] ] (NormalB (ConE (promote symbolType))) []
-        promote Regular = 'Regular
-        promote Anonymous = 'Anonymous
-        promote Auxiliary = 'Auxiliary
-        renameDups done [] = reverse done
-        renameDups done ((ty, name):queue) = if elem name (snd <$> done)
-                                      then renameDups done ((ty, name ++ "'") : queue)
-                                      else renameDups ((ty, name) : done) queue
+  let mkMatch symbolType str = match (conP (Name (OccName str) (NameQ modName)) []) (normalB [e|symbolType|]) []
+  (:) <$> dataD (pure []) name [] Nothing (flip normalC [] . mkName . snd <$> namedSymbols) [ derivClause Nothing (map conT [ ''Bounded, ''Enum, ''Eq, ''Ix, ''Ord, ''Show ]) ]
+      <*> [d|
+    instance Symbol $(conT name) where
+      symbolType = $(lamCaseE (uncurry mkMatch <$> namedSymbols)) |]
+
+renameDups :: [(a, String)] -> [(a, String)]
+renameDups = go mempty
+  where go _    []                = []
+        go done ((ty, name):rest) = let name' = if name `Set.member` done then name ++ "'" else name in (ty, name') : go (Set.insert name' done) rest
```

Comment:
This is quite a bit more straightforward, and no longer requires the `reverse`. It should also be more efficient due to the `Set`-based lookup instead of the linear search.